xdsl/dialects/pdl_interp.py

```diff
@@ -560,6 +587,110 @@ def __init__(
             },
         )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> CreateOperationOp:
+        def _parse_attribute() -> tuple[Attribute, SSAValue]:
+            attrname = parser.parse_attribute()
+            parser.parse_punctuation("=")
+            operand = parser.parse_operand()
+            return (attrname, operand)
+
+        def _parse_input_list():
+            values: list[SSAValue] = []
+            parser.parse_punctuation("(")
+            if not parser.parse_optional_punctuation(")"):
+                values = parser.parse_comma_separated_list(
+                    delimiter=Parser.Delimiter.NONE,
+                    parse=lambda: parser.parse_operand(),
+                )
+                parser.parse_punctuation(":")
+                parser.parse_comma_separated_list(
+                    delimiter=Parser.Delimiter.NONE,
+                    parse=lambda: parser.parse_type(),
+                )
+                parser.parse_punctuation(")")
+            return values
```

Comment:
this pattern redefines the helper functions each time that an operation of this class is parsed, I'd recommend moving them out to private `@static` methods on the type instead, to avoid redundant work.